File: dynamic_programming/grid/grid.cpp
/*se posicao = '.' -> caminho livre
# se posicao = '#' -> parede
# (1,1) é sempre um caminho livre
# Só se movimenta para a direita ou para baixo

# para evitar overflow usar '(a+b) % 10^9+7' ao inves de 'a+b' somente*/

//Student: Gustavo Sampaio Lima
//ID: 12623992
#include <bits/stdc++.h>
#include <iostream>
using namespace std;
#define MAXSIZE 1000

int main(){
    int rows = 0, columns = 0;
    cin >> rows >> columns;
    
    vector<vector<char>> board[MAXSIZE][MAXSIZE];
    vector<vector<int>> MEMO[MAXSIZE][MAXSIZE];

    char entry;
    for(int i = 0; i < rows; i++){
        for(int j = 0; j < columns; j++){
            cin >> entry;
            if(entry == " ") cin >> entry;
                board[i][j] = entry;
                MEMO[i][j] = -1;
        }    
    }

    return 0;
}

File: dynamic_programming/grid/grid.py
#Student: Gustavo Sampaio Lima
#ID: 12623992
from sys import setrecursionlimit
from functools import lru_cache

setrecursionlimit(5000)    
@lru_cache(maxsize=None)
def search_new_path(posY, posX):
    #out of limits
    if posY >= rows or posX >= columns: 
        return 0
    else:
        #final position
        if posY == rows-1 and posX == columns-1:
            return 1
        
        if(board[posY][posX] == '#'):
            return 0
        result = (search_new_path(posY+1, posX) + search_new_path(posY, posX+1)) % MOD
        return result 

MOD = 1000000007
rows, columns = [int(i) for i in input().split()]

board = []

for i in range(rows):
    new_column = input()
    new_column = [i for i in new_column]
    board.append(new_column)

print(search_new_path(0,0))

